pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -319,11 +319,16 @@ private boolean createDictionaryForColumn(ColumnIndexCreationInfo info, SegmentG
       return false;
     }
 
+    // Do not create dictionaries for json or text index columns as they are high-cardinality values almost always
+    if (config.getJsonIndexCreationColumns().contains(column)
+        || config.getTextIndexCreationColumns().contains(column)) {
+      return false;
+    }
+
     // Do not create dictionary if index size with dictionary is going to be larger than index size without dictionary
     // This is done to reduce the cost of dictionary for high cardinality columns
     // Off by default and needs optimizeDictionaryEnabled to be set to true
-    if (config.isOptimizeDictionaryForMetrics() && spec.getFieldType() == FieldType.METRIC && spec.isSingleValueField()
-        && spec.getDataType().isFixedWidth()) {
+    if (config.isOptimizeDictionaryForMetrics() && spec.isSingleValueField() && spec.getDataType().isFixedWidth()) {
```

Comment:
now that this config has been changed to work for both metrics and dimension fields, shouldn't `isOptimizeDictionaryForMetrics()` and the associated variable indicate that it works for both? maybe this should be a separate flag for dimension since it'll be hard to modify existing configs across all tables to change the name of this one.

In the `IndexingConfig` there is even a comment that'll need to be updated:
```
  /**
   * If `optimizeDictionaryForMetrics` enabled, dictionary is not created for the metric columns
   * for which rawIndexSize / forwardIndexSize is less than the `noDictionarySizeRatioThreshold`.
   */
```

Reply:
Yeah, it makes sense. I am planning to introduce a new config altogether even for json/text.

Reply:
On thinking about it, It will get pretty confusing for users. I have decided to finally change the config to `optimizeDictionary` and keep the old config as well for backward compatibility. Have added comments for deprecation though.